clang/test/SemaSYCL/wrong-address-taking.cpp

```diff
@@ -0,0 +1,149 @@
+// RUN: %clang_cc1 -fsycl-is-device -fsycl-allow-func-ptr -internal-isystem %S/Inputs -fsyntax-only -verify -sycl-std=2020 -std=c++17 %s
+
+#include "sycl.hpp"
+
+int badFoo(int P) {
+  return P + 2;
+}
+
+[[intel::device_indirectly_callable]] int goodFoo(int P) {
+  return P + 2;
+}
+
+SYCL_EXTERNAL float externalBadFoo(int P);
+[[intel::device_indirectly_callable]] unsigned externalGoodFoo(int P);
+
+sycl::queue myQueue;
+
+SYCL_EXTERNAL int runFn(int (&)(int));
+SYCL_EXTERNAL int runFn1(int (*)(int));
+
+struct ForMembers {
+  [[intel::device_indirectly_callable]] int goodMember(int) { return 1; }
+  int badMember(int) { return 2; }
+
+  static int badStaticMember(int) { return 2; }
+};
+
+template <typename Fn, typename... Args> void templateCaller(Fn F, Args... As) {
+  F(As...);
+}
+
+template <auto Fn, typename... Args> void templateCaller1(Args... As) {
+  // expected-error@+1 {{taking address of a function not marked with 'intel::device_indirectly_callable' attribute is not allowed in SYCL device code}}
+  Fn(As...);
+  // expected-error@+1 {{taking address of a function not marked with 'intel::device_indirectly_callable' attribute is not allowed in SYCL device code}}
+  runFn(*Fn);
+}
+
+template <auto Fn, typename... Args> void templateCaller2(Args... As) {
+  // expected-error@+1 {{taking address of a function not marked with 'intel::device_indirectly_callable' attribute is not allowed in SYCL device code}}
+  Fn(As...);
+}
```

Comment:
This test case seems duplicate of templateCaller1(). Do we need this?

Reply:
No, we can remove this. Will do.

Reply:
Done.